### ISSUE TITLE
[RFC] MaForm 

### ISSUE THREAD

**Issue by Samitier** [holaluz/margarita#419]
Hello!

Lately, trying to make a simple form in maitai I found plenty of limitations in margarita for doing form validations. So here I go!

This is all inspired by the frustrations of https://github.com/holaluz/margarita/issues/417, so maybe it's better if you take a look there before reading this.

## What is this trying to solve?

Basically, we are all using the dependency vuelidate to manage form validations. But we are using it differently depending on the project and also repeating a lot of boilerplate. I think that we should have a more streamlined solution to this.

Also, the way we are adding form validations (at least in maitai) is full of side effects, which makes the code super fragile:

```js
// a validations.js file
export default {
  couponCode: {
    isAvailableCoupon() {

      // wtf is this.coupon? this is using the scope of the component even though its in a different file
      const isAvailableCoupon = this.coupon.isAvailable

      if (!isAvailableCoupon) {

       // Why are we sending gtm events in a validation?
        this.$gtm.pushEvent({
          eventAction: 'invitations out',
          eventLabel: this.coupon.code,
        })
      }

      return isAvailableCoupon
    },
```
And redundant listeners:
```vue
<ma-text-field
      id="mgmCouponCode"
      v-model.trim="couponCode"
      class="input"
      :label="$t('submit.referral.label')"
      :placeholder="$t('submit.referral.placeholder')"
      :has-error="hasError('couponCode')"
      :error-message="getErrorMessage('couponCode')"
      @blur="validate('couponCode')"
    />
<!-- we always have to listen the blur event. Also, why hasError and validate are different functions? 
```
```js
 // and always watch this $v.$invalid prop (that is hidden by a mixin btw)
  watch: {
    '$v.$invalid': {
      handler(isInvalid) {
        if (isInvalid && this.code !== this.coupon.code) {
          this.setCoupon(ReferralCode())
        }
      },
    },
  },
```
So IMHO, we have to add a lot of boilerplate and questionable code to validate an input. Also, we have to depend on vuelidate instead of just using margarita for that. Margarita has a pseudo error managing (`error-message`/`has-error`) but it is not accessible or even usable without said boilerplate (see https://github.com/holaluz/margarita/issues/417)

## Ok, stop the rant. What do you propose?

I think we can simplify all this a lot. For a form, I think that what we all need is the ability to give a validator function to all of our inputs and also a variable that tells us if the whole form has any problem or not in order to let the user continue.

I'm going to do it with an example:

Imagine this form:

![image](https://user-images.githubusercontent.com/4160121/118618947-6d4a7000-b7c4-11eb-8d7f-6435a0b2c511.png)

It could be something like this:

```vue
<ma-form :is-form-valid.sync="isFormValid">
   <ma-text-field 
         v-model.trim="dni"
         :label="$t('dni')"
         :placeholder="$t('dni')"
         :validate="validateDni"
    />
   <ma-text-field 
         v-model.trim="name"
         :label="$t('name')"
         :placeholder="$t('name')"
         :validate="validateRequiredString"
    />
   <ma-text-field 
         v-model.trim="firstSurname"
         :label="$t('firstSurname')"
         :placeholder="$t('firstSurname')"
         :validate="validateRequiredString"
    />
   <ma-text-field 
         v-model.trim="secondSurname"
         :label="$t('secondSurname')"
         :placeholder="$t('secondSurname')"
         :validate="validateRequiredString"
    />
</ma-form>
```
A validate function could be something like:

```js
methods: {
    validateDni(dni) {
       if (!dni) return this.$t('thisFieldIsRequired')
       if (dni regex) return this.$t('invalidDni')
    },
    validateRequiredString(s) {
        if (!s) return this.$t('thisFieldIsRequired)
    }
}
```

Automatically, ma-text-field should be clever enough to validate a field on blur, without us telling it explicitly. Also to not validate anything if the field has never been touched nor has value.

Also, `ma-form` should keep a computed prop that checks whether all its child ma-text-fields are valid or not. For example, to enable/disable the submit button.

## So, do we need to introduce vuelidate in margarita or not?

I don't know. I think what I say here can be achieved without it. Do you have any use case that's different than what I'm writing here? I just know how we manage forms in dry/maitai so my view can be biased. 

Let's talk!

**Comment by edufarre** [comment]
> we always have to listen to the blur event. Also, why hasError and validate are different functions? 

The `validate` and the `hasError` functions kinda aim for the same, as you can clearly check in your `input validator mixin. Basically one triggers the validation function and the other one checks if the desired validation field does have an error... I feel your pain here, tbh I believe we could simplify this and just have one of them.

> So IMHO, we have to add a lot of boilerplate and questionable code to validate the input. Also, we have to depend on vuelidate instead of just using margarita for that.

You are right, we basically started using this library because I found that this one the best approach two years ago when working on Bonasera. However, it does require a lot of boilerplate to handle the errors (like emitting or passing by prop if the current field is errored or not 🤷🏽 ).

> Automatically, ma-text-field should be clever enough to validate a field on blur, without us telling it explicitly. Also to not validate anything if the field has never been touched nor has value.

I personally think that this does depend of the form use case, depending on what and how we do expect it to behave. Moreover, its behaviour should be the same in ALL the different forms we implement.

> Do you have any use case that's different than what I'm writing here? 

As I said, we are using `vuelidate` cuz before we started working on Bonasera we had no library on the frontend to validate the different inputs and this lib looked like it was more than enough. We are currently handling lots of forms and business logic in them and I reckon we could try to improve this.

I would NOT go for the approach you suggested cuz basically these validation libraries do handle lots of things for us already, like some common validations _(is a number, is a string, required, ...)_ and helpful props.
With the Power Rangers team, we have taken another approach on `Lumo360` (you can find the same kinda approach on `nuxt-clean`), we swapped the validation library in order to use [VeeValidate](https://vee-validate.logaretm.com/v3) which provides us with all the needed requirements we had.
This lib allows us to have a [generic file](https://github.com/holaluz/lumo360/blob/master/frontend/src/ui/plugins/veeValidate.ts) where we do have ALL the different validations (the ones that come with the library and the custom ones, thus being able to re-use them over and over). Using `VeeValidate` in our `.vue` files is really easy thanks to the [ValidationProvider](https://vee-validate.logaretm.com/v3/api/validation-provider.html) and the [ValidationObserver](https://vee-validate.logaretm.com/v3/api/validation-observer.html) components. We basically created a [wrapper component](https://github.com/holaluz/lumo360/blob/master/frontend/src/ui/components/form-element-wrapper/form-element-wrapper.vue) that allows us to handle the form errors (including the fields that have more than one error 😮 🎉 ) and all the different validations in an easy way in each of the desired form templates like [this validation case](https://github.com/holaluz/lumo360/blob/bf63929db2d78a2505dde9d72337f4ea37a25f7e/frontend/src/ui/pages/power-change/components/power-change-form.vue#L22).

I won't get into the documentation, but using `VeeValidate` we realised that it lots of cool functionalities to offer to us.

In a nutshell, I do not think we should implement our own `form validation lib` because looks like we are 're-inventing the wheel' when they already exist, in my opinion, I would go for a library and stick to it, one we all find useful, easy to use and functional. And if I have to vote, I would stay with `VeeValidate` instead of `vuelidate` even though that would require updating all the other repos and their validations, right? (sounds like lots of work)

**Comment by Samitier** [comment]
Aww yeah, thanks for the overview, I didn't know we were using another validation library. This looks better than vuelidate tbh (and more vue 3 friendly 😂 ).

> In a nutshell, I do not think we should implement our own form validation lib

I completely aggree, if we can avoid that! What we can do, though, is incorporate a validation library in margarita so that we implement all of our forms the same way. VeeValidate seems like a very nice, composable solution that could even be applied with the API proposed in this RFC.

For example `ma-form` could be a wrapper for a `validation-observer`. We could also have the basic validations (required, string, number) shared between projects. And even share validators for more complex use cases (like cups, postal codes and DNI/NIF/IBANS in a repo like `pomada`). The nice thing is that if we use a validation library, we are constrained to implement all of our validators the same way, which can make our shared code way more cohesive.

I have a question in one of the snippets you provided. How do you show the error message under an input if the validation fails? Seems like we are not using margarita's `has-error` or `error-message` here: 

```html
<form-element-wrapper rules="required|numeric">
          <ma-select
                v-if="shouldDisplayNormalizedPowersSelect"
                v-model.number="power.amount"
                :options="tariff.normalizedPowers"
                :formatter="normalizePowerFormatter"
                :label="`${$t('power')} ${power.period}`"
          />
          <ma-text-field
                v-else
                v-model.number="power.amount"
                :label="`${$t('power')} ${power.period}`"
          />
</form-element-wrapper>
```
How would you incorporate margarita's error messages with veeValidate? Do you think it's easy?

> And if I have to vote, I would stay with VeeValidate instead of vuelidate even though that would require updating all the other repos and their validations, right? (sounds like lots of work)

If we don't incorporate it in margarita I wouldn't change dependency in the "old" repos. Seems like a bit of a wasted time. If we do, yeah, we will have to change all of them if we are not careful and make a breaking change. Seems like it could easily get out of hand if we have to make everyone redo their forms. Maybe we should think first if this is viable or not. Or just make an entire new component and not touch `ma-text-field` so that it could be "optional" to change to veeValidate.

**Comment by Samitier** [comment]
closed by https://github.com/holaluz/kalimotxo/pull/86
